Handle shorthands in attributes correctly

ezhlobo opened the pull request:
It adds supporting of shorthands in attributes. Initial issue [#1 (Fails on boolean attributes)](https://github.com/pugjs/babel-plugin-transform-react-pug/issues/1).

Now following syntax won't throw an error:

```pug
div(first second third)
```

But this one will throw:

```pug
div(
  first!="stirng" 
  second!=variable
)
```

Checklist:

- [x] Use `context.error` instead of `new Error` to keep well-described errors
- [x] Update `pug-lexer` and remove workaround to keep `mustEscape` correct

ezhlobo commented:
The main reason why I created this PR is to get feedback about my direction.

ForbesLindesay commented on src/visitors/Tag.js:
```diff
@@ -57,8 +67,8 @@ function getAttributes(
 
       const expr = parseExpression(val === true ? 'true' : val, context);
 
-      if (!mustEscape && (!t.isStringLiteral(expr) || /(\<\>\&)/.test(val))) {
```
Why are you removing the `(!t.isStringLiteral(expr) || /(\<\>\&)/.test(val))`?

ezhlobo replied:
@ForbesLindesay because it looks impossible to achieve with basic tests. Maybe we can hack it somehow... Otherwise `val` starts and ends with `"` when it's a string. 

And filtering off `<>&` looks not obvious for me, why not `<script>...` or something like that? Probably we lost `g` flag for regex?

Anyway I'm going to use different conditions. All use cases are captured in tests. Let's discuss something, if you want to add to those tests.

ForbesLindesay replied:
The case this was capturing, although arguably not very important, is:

```pug
div(data-whatever!="some string that does not contain any special characters")
```

The idea being that it might make it slightly easier when converting from html based pug templates to react pug templates.  Arguably that might not be super important though.

ForbesLindesay commented on src/__tests__/attributes-shorthand.input.js:
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+// To prevent warnings in console from react
+const Custom = () => null
```
You could also render an `input(checked)` so we can actually see html output in the test.

ezhlobo replied:
@ForbesLindesay I started using `data-` attributes to keep html output. Let me know if you think it does not appropriate.

ForbesLindesay commented:
Looks good. Ping me once you've fixed build & made the changes you want to make.

ezhlobo commented:
@ForbesLindesay I rewrote my solution completely (left only test cases), need your attention one more time.

Also it looks like I need your help with publishing new version of `pug-lexer` (after [#2956](https://github.com/pugjs/pug/pull/2956) and [#2957](https://github.com/pugjs/pug/pull/2957)). Otherwise we need to support a workaround in our codebase.

ForbesLindesay commented on src/visitors/Tag.js:
```diff
@@ -54,8 +69,19 @@ function getAttributes(
-        throw new Error('Unescaped attributes are not supported in react-pug');
+      if (!mustEscape) {
+        const isStringViaAliases: boolean =
+          t.isStringLiteral(expr) && !['className', 'id'].includes(name);
```
For a list of two, I think I'd prefer just `(name === 'className' || name === 'id')` - it takes less time to mentally parse.

ForbesLindesay commented on src/visitors/Tag.js:
```diff
@@ -54,8 +69,19 @@ function getAttributes(
-      if (!mustEscape && (!t.isStringLiteral(expr) || /(\<\>\&)/.test(val))) {
-        throw new Error('Unescaped attributes are not supported in react-pug');
+      if (!mustEscape) {
+        const isStringViaAliases: boolean =
```
You shouldn't need to explicitly add the `: boolean` here; flow should infer it from context.

ForbesLindesay commented on src/visitors/Tag.js:
```diff
@@ -54,8 +69,19 @@ function getAttributes(
+        const isStringViaAliases: boolean =
+          t.isStringLiteral(expr) && !['className', 'id'].includes(name);
+
+        const isNotStringOrBoolean: boolean =
```
ditto

ForbesLindesay commented:
I don't have a strong opinion on whether we should support `div(data-whatever!="some string that does not contain any special characters")` or not.

Please can you address the other small comments though.

ezhlobo commented:
@ForbesLindesay thank you for your review. I addressed all your comments and removed my workaround because there is new version of pug-lexer (that made it for us).

The question about checking tags in scripts is still open. I would suggest to create a separate issue and discuss it there, because anyway it did not work as expected before this PR. Does it work for you? We can actually remember all use cases of unescaped values and think up how to handle them best...

ForbesLindesay commented on src/visitors/Tag.js:
```diff
@@ -54,8 +59,19 @@ function getAttributes(
+          t.isStringLiteral(expr) && name !== 'className' && name !== 'id';
+
+        const isNotStringOrBoolean =
+          !t.isStringLiteral(expr) && !t.isBooleanLiteral(expr);
```
I'm just wondering if this lot can be clarified. It seems like the allowed case is just:

```js
t.isBooleanLiteral(expr) ||
((name === 'className' || name === 'id') && t.isStringLiteral(expr))
```

It seems simpler written like that to me, rather than having multiple references to `isStringLiteral`.  How about we just use:

```js
const isAllowedUnescapedAttribute = (
  t.isBooleanLiteral(expr) ||
  ((name === 'className' || name === 'id') && t.isStringLiteral(expr))
);
if (!isAllowedUnescapedAttribute) {
```

ezhlobo replied:
@ForbesLindesay I agree that my naming can confuse, but it was created with following reason in mind: all not obvious conditions should be commented. Variables are more convenient than just comments, so I used them.

`isAllowedUnescapedAttribute` sounds not obvious because we don't use such naming through this plugin and pug (I mean it does not respect spirit of `mustEscape`). It's better to keep `shouldBeEscaped`, `allowedToBeEscaped`.

### Let me show how did I achieve my code:

When attribute is marked that we can skip its escaping, we should always show a message about unsupported feature. Two exceptions, where we should not show a message: 

1. if attribute is a string not from aliases (not from `className` and `id`)
2. if attribute is not primitive (in pug it's everything besides string and booleans)

If I convert it to the code:

```js
if (!mustEscape) {
  const isStringNotFromAliases = 
    t.isStringLiteral(expr) && name !== 'className' && name !== 'id';

  const isNotPrimitive =
    !t.isStringLiteral(expr) && !t.isBooleanLiteral(expr);

  if (isStringNotFromAliases || isNotPrimitive) {
    throw context.error(
      'INVALID_EXPRESSION',
      'Unescaped attributes are not supported in react-pug',
    );
  }
}
```

This snippet is what I suggest to leave. But I don't mind to make it much more compact by refactoring javascript, not the logic. If this is the case, then could you suggest any concrete solution that I can use? Maybe you can help me with optimizing the logic?

ForbesLindesay replied:
I'd much rather see a whitelist of cases where it is ok to have unescaped attributes:

1. `className` and `id` if the value is a string literal
2. any attribute if the value is a boolean

As this is much easier for me to interpret. When I read your code, I have to mentally figure out which are the allowed cases. Since the allowed cases are so narrow, I never really feel like I need to figure out what the non-allowed cases are.

`allowedToBeEscaped` doesn't make sense because everything *can* be escaped. `mustEscape` is fine, but we already have that with a subtly different meaning. I think something like `escapingNotRequired` or `canSkipEscaping` makes more sense here.

ezhlobo replied:
@ForbesLindesay you are right that understanding whitelist is much better here. But I still need your help with figuring out the right way. Does following code look better:

```js
if (!mustEscape) {
  const canSkipEscaping =
    (name === 'className' || name === 'id') && t.isStringLiteral(expr);

  if (!canSkipEscaping) {
    throw context.error(
      'INVALID_EXPRESSION',
      'Unescaped attributes are not supported in react-pug',
    );
  }
}
```

I noticed that we should not check is the value boolean or not. We don't allow to write `!=` constructions at all. Could you please take a look at tests in this PR and tell me is everything appropriate there?

ForbesLindesay replied:
Yes, although you've dropped the `BooleanLiteral` case there. Is that inentional?

ForbesLindesay commented:
I'm not sure if it worked for me or not, I haven't actually made use of that bit anyway.

ezhlobo commented:
@ForbesLindesay hey, I pushed updates according to our discussion. Need your attention one more time.

>you've dropped the BooleanLiteral case there. Is that intentional?

Yes. Now all boolean attributes come to us with `mustEscape: true`, so it is not necessary for us.

But the following example will throw an error and it's intentionally:

```pug
div(name!=true)
```

ForbesLindesay commented:
Thanks :)

ezhlobo commented:
Thank you very much for your time and patience
